compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java

```diff
@@ -3869,28 +3876,71 @@ public void visit(BLangOnFailClause onFailClause, AnalyzerData data) {
         // Create a new block environment for the on-fail node.
         SymbolEnv onFailEnv = SymbolEnv.createBlockEnv(onFailClause.body, data.env);
         VariableDefinitionNode onFailVarDefNode = onFailClause.variableDefinitionNode;
+        Stack<LinkedHashSet<BType>> onFailErrTypes = data.commonAnalyzerData.errorTypes;
+
         if (onFailVarDefNode != null) {
+            BLangVariable variableNode = (BLangVariable) onFailVarDefNode.getVariable();
+            if (variableNode.getKind() != NodeKind.VARIABLE && variableNode.getKind() != NodeKind.ERROR_VARIABLE) {
+                dlog.error(variableNode.pos, DiagnosticErrorCode.INVALID_BINDING_PATTERN_IN_ON_FAIL);
+            }
+
+            BType failErrorType;
+            LinkedHashSet<BType> onFailErrType = onFailErrTypes.peek();
+            boolean isVarTypeHardcoded = false;
+            if (onFailErrType.size() == 1) {
+                failErrorType = onFailErrType.iterator().next();
+            } else if (onFailErrType.size() > 1) {
+                failErrorType = BUnionType.create(null, onFailErrType);
+            } else {
+                failErrorType = symTable.errorType;
+                isVarTypeHardcoded = true;
+            }
+
             // Check on-fail node's variables and set types.
-            handleForeachDefinitionVariables(onFailVarDefNode, symTable.errorType,
-                    onFailClause.isDeclaredWithVar, true, onFailEnv);
+            handleForeachDefinitionVariables(onFailVarDefNode, failErrorType,
+                    onFailClause.isDeclaredWithVar, isVarTypeHardcoded, onFailEnv);
             BLangVariable onFailVarNode = (BLangVariable) onFailVarDefNode.getVariable();
-            if (!types.isAssignable(onFailVarNode.getBType(), symTable.errorType)) {
+
+            if (onFailVarNode.getBType() != null && !types.isAssignable(onFailVarNode.getBType(), symTable.errorType)) {
                 dlog.error(onFailVarNode.pos, DiagnosticErrorCode.INVALID_TYPE_DEFINITION_FOR_ERROR_VAR,
                         onFailVarNode.getBType());
             }
+
+            for (BType errorType : onFailErrTypes.peek()) {
```

Comment:
Doesn't `handleForeachDefinitionVariables` already log errors when incompatible?

The following results in 3 errors.

```ballerina
public function main() {
    do {
        if true {
            fail error E1("e1", j = 1);
        }
        fail error E2("e2", k = 2);
    } on fail Error error(i = i) {

    }
}

type Error error<record {| int i; |}>;
type E1 error<record {| int j; |}>;
type E2 error<record {| int k; |}>;
```

```
ERROR [x.bal:(7:15,7:33)] incompatible error definition type: 'E1' will not be matched to 'Error'
ERROR [x.bal:(7:15,7:33)] incompatible error definition type: 'E2' will not be matched to 'Error'
ERROR [x.bal:(7:15,7:20)] incompatible types: expected '(E1|E2)', found 'Error'
```